File: 160_intersection_of_two_linked_lists/intersection_of_two_linked_lists.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

    def __eq__(self, other):
        # 注意，为了简化判断结果，假设节点值不重复
        return self.val == other.val


def list_to_node(nums: list) -> ListNode:
    """从列表构建链表"""
    dummy = ListNode(None)
    node = dummy
    for n in nums:
        node.next = ListNode(n)
        node = node.next
    return dummy.next


class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        l1, l2 = [], []
        while headA:
            l1.append(headA)
            if headA.next:
                headA = headA.next
            else:
                break
        while headB:
            l2.append(headB)
            if headB.next:
                headB = headB.next
            else:
                break
        ret = None
        while l1 and l2 and (va := l1.pop()) == (vb := l2.pop()):
            ret = va
            continue
        return ret

File: 160_intersection_of_two_linked_lists/intersection_of_two_linked_lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode GetIntersectionNode(ListNode headA, ListNode headB) {
        Stack<ListNode> l1 = new Stack<ListNode>();
        Stack<ListNode> l2 = new Stack<ListNode>();
        while (headA != null)
        {
            l1.Push(headA);
            headA = headA.next;
        }
        while (headB != null)
        {
            l2.Push(headB);
            headB = headB.next;
        }

        ListNode ret = null;
        while (l1.Count() > 0 && l2.Count() > 0)
        {
            ListNode tmp = l1.Pop();
            if (tmp == l2.Pop())
                ret = tmp;
            else
                break;
        }
        return ret;
    }
}